search/process/transform.py

```diff
@@ -0,0 +1,165 @@
+"""Responsible for transforming metadata & fulltext into a search document."""
+
+from datetime import datetime
+from string import punctuation
+from typing import Callable, Dict, List, Optional, Tuple, Union
+from search.domain import Document, DocMeta, Fulltext
+
+
+DEFAULT_LICENSE = {
+    'uri': 'http://arxiv.org/licenses/assumed-1991-2003/',
+    'label': "Assumed arXiv.org perpetual, non-exclusive license to distribute"
+             " this article for submissions made before January 2004"
+}
+
+
+def _constructLicense(meta: DocMeta) -> dict:
+    """Get the document license, or set the default."""
+    if not meta.license or not meta.license['uri']:
+        return DEFAULT_LICENSE
+    return meta.license
+
+
+def _strip_punctuation(s: str) -> str:
+    return ''.join([c for c in s if c not in punctuation])
+
+
+def _constructPaperVersion(meta: DocMeta) -> str:
+    """Generate a version-qualified paper ID."""
+    if 'v' in meta.paper_id:
+        return meta.paper_id
+    return '%sv%i' % (meta.paper_id, meta.version)
+
+
+def _constructMSCClass(meta: DocMeta) -> Optional[list]:
+    """Extract ``msc_class`` field as an array."""
+    if not meta.msc_class:
+        return None
+    return [obj.strip() for obj in meta.msc_class.split(',')]
+
+
+def _constructACMClass(meta: DocMeta) -> Optional[list]:
+    """Extract ``acm_class`` field as an array."""
+    if not meta.acm_class:
+        return None
+    return [obj.strip() for obj in meta.acm_class.split(';')]
+
+
+def _transformAuthor(author: dict) -> dict:
+    fname = _strip_punctuation(author['first_name'])
+    author['initials'] = [pt[0] for pt in fname.split() if pt]
+    author['full_name'] = f"{author['first_name']} {author['last_name']}"
+    return author
+
+
+def _constructAuthors(meta: DocMeta) -> List[Dict]:
+    return [_transformAuthor(author) for author in meta.authors_parsed]
+
+
+def _constructAuthorOwners(meta: DocMeta) -> List[Dict]:
+    return [_transformAuthor(author) for author in meta.author_owners]
+
+
+def _getFirstSubDate(meta: DocMeta) -> Optional[str]:
+    if not meta.submitted_date_all:
+        return None
+    return meta.submitted_date_all[0]
+
+
+def _getLastSubDate(meta: DocMeta) -> Optional[str]:
+    if not meta.submitted_date_all:
+        return None
+    return meta.submitted_date_all[-1]
+
+
+TransformType = Union[str, Callable]
+_transformations: List[Tuple[str, TransformType]] = [
+    ('id', 'paper_id'),
+    ('abstract', 'abstract'),
+    ('authors', _constructAuthors),
+    ('authors_freeform', "authors_utf8"),
+    ("owners", _constructAuthorOwners),
```

Comment:
Mixed quote types. Can we have just one or the other? Can we add that to linting?

Reply:
Thanks, we'll make this consistent.